### PR TITLE
Bug 1724053. Retrieve oauthclient as a cluster resource

### PR DESCRIPTION
**Pull request by jcantrill**
Fixes:
https://bugzilla.redhat.com/show_bug.cgi?id=1724053

The reconcile process of updating oauthclient should GET a cluster level resource, not namespaced

### CODE REVIEW

File: pkg/k8shandler/clusterloggingrequest.go [hunk 1]
```diff
@@ -35,6 +35,12 @@ func (clusterRequest *ClusterLoggingRequest) Get(objectName string, object runti
 	return clusterRequest.client.Get(context.TODO(), namespacedName, object)
 }
 
+func (clusterRequest *ClusterLoggingRequest) GetClusterResource(objectName string, object runtime.Object) error {
+	namespacedName := types.NamespacedName{Name: objectName}
+
+	return clusterRequest.client.Get(context.TODO(), namespacedName, object)
```
Comment (richm) [inline pkg/k8shandler/clusterloggingrequest.go]: Are there docs that say that you omit the namespace name from the object to get a cluster scoped object?

**Comment by jcantrill** [comment]
> LGTM. Please remove the debug statements before merging

Why remove the debug statements?  They are only logged when the loglevel is set to debug

File: pkg/k8shandler/visualization.go [hunk 2]
```diff
@@ -240,6 +243,7 @@ func (clusterRequest *ClusterLoggingRequest) createOrUpdateOauthClient(oauthSecr
 			current.Secret = oauthClient.Secret
 			current.ScopeRestrictions = oauthClient.ScopeRestrictions
 			if err = clusterRequest.Update(current); err != nil {
+				logrus.Errorf("Error Updating oauthclient %q: %v", oauthClient.Name, err)
```
Comment (ewolinetz) [inline pkg/k8shandler/visualization.go]: fyi this error should bubble all the way up and be printed already

File: pkg/k8shandler/visualization.go [hunk 3]
```diff
@@ -197,6 +197,7 @@ func (clusterRequest *ClusterLoggingRequest) createOrUpdateOauthClient(oauthSecr
 
 	kibanaURL, err := clusterRequest.GetRouteURL("kibana")
 	if err != nil {
+		logrus.Errorf("Error fetching kibana route %v", err)
```
Comment (ewolinetz) [inline pkg/k8shandler/visualization.go]: this should bubble all the way up already...

File: pkg/k8shandler/visualization.go [hunk 4]
```diff
@@ -219,6 +220,7 @@ func (clusterRequest *ClusterLoggingRequest) createOrUpdateOauthClient(oauthSecr
 	utils.AddOwnerRefToObject(oauthClient, utils.AsOwner(clusterRequest.cluster))
 
 	err = clusterRequest.Create(oauthClient)
+	logrus.Debugf("Response %v", err)
```
Comment (ewolinetz) [inline pkg/k8shandler/visualization.go]: can we move this debug into the `Create` function instead?

File: pkg/k8shandler/visualization.go [hunk 5]
```diff
@@ -227,10 +229,11 @@ func (clusterRequest *ClusterLoggingRequest) createOrUpdateOauthClient(oauthSecr
 		current := oauthClient.DeepCopy()
 
 		retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
-			if err = clusterRequest.Get(oauthClient.Name, current); err != nil {
+			if err = clusterRequest.GetClusterResource(oauthClient.Name, current); err != nil {
 				if errors.IsNotFound(err) {
 					// the object doesn't exist -- it was likely culled
 					// recreate it on the next time through if necessary
+					logrus.Debugf("Returning nil. The oauthclient %q was not found even though create previously failed.  Was it culled?", oauthClient.Name)
```
Comment (ewolinetz) [inline pkg/k8shandler/visualization.go]: can we move this debug into `GetClusterResource` instead?

File: pkg/k8shandler/visualization.go [hunk 6]
```diff
@@ -188,7 +188,7 @@ func (clusterRequest *ClusterLoggingRequest) createOrUpdateKibanaDeployment() (e
 }
 
 func (clusterRequest *ClusterLoggingRequest) createOrUpdateOauthClient(oauthSecret string) (err error) {
-
+	logrus.Debug("createOrUpdateOauthClient for Kibana...")
```
Comment (ewolinetz) [inline pkg/k8shandler/visualization.go]: not sure this buys us anything, it should get into this function every time it processes unless a preceding one returns an error

File: pkg/k8shandler/clusterloggingrequest.go (hunk 1 above)
Comment (jcantrill) [reply pkg/k8shandler/clusterloggingrequest.go]: I didnt find any.  i simply guessed

Comment (richm) [reply pkg/k8shandler/clusterloggingrequest.go]: ok - we'll know soon enough if it works - but something to keep in mind that we are using an unknown and/or undocumented api feature that may break unexpectedly in the future

File: pkg/k8shandler/visualization.go [hunk 7]
```diff
@@ -165,6 +167,9 @@ func (clusterRequest *ClusterLoggingRequest) createOrUpdateKibanaDeployment() (e
 		"kibana",
 		kibanaPodSpec,
 	)
+	kibanaDeployment.ObjectMeta.Annotations = map[string]string{
```
Comment (ewolinetz) [inline pkg/k8shandler/visualization.go]: this logic isn't enough, we don't check annotation values at all while determining if an update is required (in `isKibanaDifferent`)

File: pkg/k8shandler/clusterloggingrequest.go (hunk 1 above)
Comment (ewolinetz) [reply pkg/k8shandler/clusterloggingrequest.go]: @jcantrill did you see it came back with the correct object during your testing with `debug` statements?

File: pkg/k8shandler/visualization.go [hunk 8]
```diff
@@ -45,16 +50,12 @@ func (clusterRequest *ClusterLoggingRequest) CreateOrUpdateVisualization() (err
 			oauthSecret = utils.GetRandomWord(64)
 			utils.WriteToWorkingDirFile("kibana-proxy-oauth.secret", oauthSecret)
 		}
-
-		if err = clusterRequest.createOrUpdateKibanaSecret(oauthSecret); err != nil {
```
Comment (ewolinetz) [inline pkg/k8shandler/visualization.go]: is there a reason you removed this? I would expect that we don't see the secret created at all now...

File: pkg/k8shandler/visualization.go (hunk 7 above)
Comment (ewolinetz) [reply pkg/k8shandler/visualization.go]: This is still missing

Comment (jcantrill) [reply pkg/k8shandler/visualization.go]: Does this not resolve your ask: https://github.com/openshift/cluster-logging-operator/pull/217/files#diff-5ff1cbe659b99e0e73d8ba484249c27cR579

File: pkg/k8shandler/visualization.go (hunk 8 above)
Comment (jcantrill) [reply pkg/k8shandler/visualization.go]: I dont think this was intentional... may be a bad merge or the way I brought in the file

File: pkg/k8shandler/visualization.go (hunk 7 above)
Comment (ewolinetz) [reply pkg/k8shandler/visualization.go]: i was thrown off by the function name... can we abstract the logic in `AreSelectorsSame` to something like `areMapsSame` and have a `AreSelectorsSame` and `AreAnnotationsSame` and call `areMapsSame` underneath that?

Comment (jcantrill) [reply pkg/k8shandler/visualization.go]: @ewolinetz are you agreeable to skipping the layering and just using `AreMapsSame`?  Is there a reason we cannot use `reflect.DeepEquals` and remove the function all together?

Comment (jcantrill) [reply pkg/k8shandler/visualization.go]: reflect is the recommended solution: https://stackoverflow.com/questions/18208394/how-to-test-the-equivalence-of-maps-in-golang

Comment (richm) [reply pkg/k8shandler/visualization.go]: I'm using `reflect.DeepEqual` in the rsyslog undefined_field tests: https://github.com/openshift/origin-aggregated-logging/blob/master/rsyslog/undefined_field/undefined_field_test.go#L279

Comment (ewolinetz) [reply pkg/k8shandler/visualization.go]: yeah, i had tried to use `reflect` before but found it didn't correctly evaluate which is why its a function instead... :man_shrugging:

File: pkg/k8shandler/visualization.go [hunk 9]
```diff
@@ -236,20 +240,29 @@ func (clusterRequest *ClusterLoggingRequest) createOrUpdateOauthClient(oauthSecr
 				return fmt.Errorf("Failed to get %v oauthclient for %q: %v", oauthClient.Name, clusterRequest.cluster.Name, err)
 			}
 
+			//Check to see if we even need to update
+			//remove annotation to normalize
+			updatedAt, _ = current.GetObjectMeta().GetAnnotations()[annotationOauthSecretUpdatedAt]
+			delete(current.GetObjectMeta().GetAnnotations(), annotationOauthSecretUpdatedAt)
```
Comment (ewolinetz) [inline pkg/k8shandler/visualization.go]: instead of doing this could we not just check `RedirectUIRs` `Secret` and `ScopeRestrictions` ?

File: pkg/utils/utils_test.go [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+	}
+	two := map[string]string{
+		"xyz": "123",
+		"foo": "bar",
+	}
+	if !AreMapsSame(one, two) {
+		t.Errorf("Exp maps to evaluate to be equivalent - left: %v, right: %v", one, two)
+	}
+}
+func TestAreMapsSameWhenDifferent(t *testing.T) {
```
Comment (ewolinetz) [inline pkg/utils/utils_test.go]: Can we also test comparing two empty maps as well as an empty map compared to a non-empty map?

File: pkg/k8shandler/visualization.go (hunk 9 above)
Comment (jcantrill) [reply pkg/k8shandler/visualization.go]: Isn't it easier to simply delete the thing we know to be different?  Also, why do we copy oauthClient in the first place if the 'Get' call is going to load the object anyway?

Comment (ewolinetz) [reply pkg/k8shandler/visualization.go]: we don't need to copy it.... either copy or create an empty oauthclient object. the copy method is left over from the older sdk logic we used...

Comment (ewolinetz) [reply pkg/k8shandler/visualization.go]: right but if the updatedAt field is different we aren't going to do an update. we only care to check if those three fields i listed above are different... so lets just check those.

**Comment by ewolinetz** [comment]
@jcantrill the failure is regarding kibana, can you please run your test locally i don't think this is a flake

**Comment by ewolinetz** [comment]
> @jcantrill the failure is regarding kibana, can you please run your test locally i don't think this is a flake

Disregard :)

File: pkg/k8shandler/visualization.go [hunk 11]
```diff
@@ -236,20 +241,26 @@ func (clusterRequest *ClusterLoggingRequest) createOrUpdateOauthClient(oauthSecr
...
+			//Check to see if we even need to update
+			if areOAuthClientsSame(current, oauthClient) {
+				return nil
+			}
+
+			updatedAt = time.Now().String()
 			current.RedirectURIs = oauthClient.RedirectURIs
 			current.Secret = oauthClient.Secret
 			current.ScopeRestrictions = oauthClient.ScopeRestrictions
+			current.GetObjectMeta().GetAnnotations()[annotationOauthSecretUpdatedAt] = updatedAt
```
Comment (richm) [inline pkg/k8shandler/visualization.go]: Just got a crash at this line:
```
E0730 22:20:34.279519       1 runtime.go:69] Observed a panic: "assignment to entry in nil map" (assignment to entry in nil map)
/go/src/github.com/openshift/cluster-logging-operator/_output/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go:76
/go/src/github.com/openshift/cluster-logging-operator/_output/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
/go/src/github.com/openshift/cluster-logging-operator/_output/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
/usr/local/go/src/runtime/asm_amd64.s:573
/usr/local/go/src/runtime/panic.go:502
/usr/local/go/src/runtime/hashmap_fast.go:696
/go/src/github.com/openshift/cluster-logging-operator/_output/src/github.com/openshift/cluster-logging-operator/pkg/k8shandler/visualization.go:253
```
The rest of the stacktrace is included below but the crash is at this line - either `current.GetObjectMeta()` or `current.GetObjectMeta().GetAnnotations()` is `nil`

```
/go/src/github.com/openshift/cluster-logging-operator/_output/src/k8s.io/client-go/util/retry/util.go:64
/go/src/github.com/openshift/cluster-logging-operator/_output/src/k8s.io/apimachinery/pkg/util/wait/wait.go:203
/go/src/github.com/openshift/cluster-logging-operator/_output/src/k8s.io/client-go/util/retry/util.go:63
/go/src/github.com/openshift/cluster-logging-operator/_output/src/github.com/openshift/cluster-logging-operator/pkg/k8shandler/visualization.go:234
/go/src/github.com/openshift/cluster-logging-operator/_output/src/github.com/openshift/cluster-logging-operator/pkg/k8shandler/visualization.go:60
/go/src/github.com/openshift/cluster-logging-operator/_output/src/github.com/openshift/cluster-logging-operator/pkg/k8shandler/reconciler.go:28
/go/src/github.com/openshift/cluster-logging-operator/_output/src/github.com/openshift/cluster-logging-operator/pkg/controller/clusterlogging/clusterlogging_controller.go:100
/go/src/github.com/openshift/cluster-logging-operator/_output/src/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:215
/go/src/github.com/openshift/cluster-logging-operator/_output/src/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158
/go/src/github.com/openshift/cluster-logging-operator/_output/src/k8s.io/apimachinery/pkg/util/wait/wait.go:133
/go/src/github.com/openshift/cluster-logging-operator/_output/src/k8s.io/apimachinery/pkg/util/wait/wait.go:134
/go/src/github.com/openshift/cluster-logging-operator/_output/src/k8s.io/apimachinery/pkg/util/wait/wait.go:88
/usr/local/go/src/runtime/asm_amd64.s:2361
panic: assignment to entry in nil map [recovered]
	panic: assignment to entry in nil map

goroutine 507 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/src/github.com/openshift/cluster-logging-operator/_output/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x107
panic(0x13b83c0, 0x16b0520)
	/usr/local/go/src/runtime/panic.go:502 +0x229
github.com/openshift/cluster-logging-operator/pkg/k8shandler.(*ClusterLoggingRequest).createOrUpdateOauthClient.func1(0x1f2c520, 0x411b09)
	/go/src/github.com/openshift/cluster-logging-operator/_output/src/github.com/openshift/cluster-logging-operator/pkg/k8shandler/visualization.go:253 +0x36f
k8s.io/client-go/util/retry.RetryOnConflict.func1(0xc420c49888, 0x412358, 0x20)
	/go/src/github.com/openshift/cluster-logging-operator/_output/src/k8s.io/client-go/util/retry/util.go:64 +0x36
k8s.io/apimachinery/pkg/util/wait.ExponentialBackoff(0x989680, 0x3ff0000000000000, 0x3fb999999999999a, 0x5, 0xc420e2a740, 0x412358, 0x30)
	/go/src/github.com/openshift/cluster-logging-operator/_output/src/k8s.io/apimachinery/pkg/util/wait/wait.go:203 +0x9c
k8s.io/client-go/util/retry.RetryOnConflict(0x989680, 0x3ff0000000000000, 0x3fb999999999999a, 0x5, 0xc420dca4b0, 0xe, 0xc42059f460)
	/go/src/github.com/openshift/cluster-logging-operator/_output/src/k8s.io/client-go/util/retry/util.go:63 +0xba
github.com/openshift/cluster-logging-operator/pkg/k8shandler.(*ClusterLoggingRequest).createOrUpdateOauthClient(0xc42070ed00, 0xc420043280, 0x40, 0x40, 0xc420043280, 0x40, 0x0)
	/go/src/github.com/openshift/cluster-logging-operator/_output/src/github.com/openshift/cluster-logging-operator/pkg/k8shandler/visualization.go:234 +0x656
github.com/openshift/cluster-logging-operator/pkg/k8shandler.(*ClusterLoggingRequest).CreateOrUpdateVisualization(0xc42070ed00, 0x0, 0x0)
	/go/src/github.com/openshift/cluster-logging-operator/_output/src/github.com/openshift/cluster-logging-operator/pkg/k8shandler/visualization.go:60 +0x1f0
github.com/openshift/cluster-logging-operator/pkg/k8shandler.Reconcile(0xc420a3a000, 0x16da3e0, 0xc4206e3bf0, 0xc420789fe0, 0x11)
	/go/src/github.com/openshift/cluster-logging-operator/_output/src/github.com/openshift/cluster-logging-operator/pkg/k8shandler/reconciler.go:28 +0x27b
github.com/openshift/cluster-logging-operator/pkg/controller/clusterlogging.(*ReconcileClusterLogging).Reconcile(0xc420704560, 0xc420789fe0, 0x11, 0xc4208fde50, 0x8, 0x1f2b180, 0x3, 0x0, 0x3)
	/go/src/github.com/openshift/cluster-logging-operator/_output/src/github.com/openshift/cluster-logging-operator/pkg/controller/clusterlogging/clusterlogging_controller.go:100 +0x167
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc4209705a0, 0x0)
	/go/src/github.com/openshift/cluster-logging-operator/_output/src/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:215 +0x188
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1()
	/go/src/github.com/openshift/cluster-logging-operator/_output/src/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158 +0x36
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc4206fe3b0)
	/go/src/github.com/openshift/cluster-logging-operator/_output/src/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x54
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc4206fe3b0, 0x3b9aca00, 0x0, 0x1, 0xc42099aa20)
	/go/src/github.com/openshift/cluster-logging-operator/_output/src/k8s.io/apimachinery/pkg/util/wait/wait.go:134 +0xbd
k8s.io/apimachinery/pkg/util/wait.Until(0xc4206fe3b0, 0x3b9aca00, 0xc42099aa20)
	/go/src/github.com/openshift/cluster-logging-operator/_output/src/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x4d
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start
	/go/src/github.com/openshift/cluster-logging-operator/_output/src/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:157 +0x35b
```
